### ISSUE TITLE
Implement per player digging speed multiplier for groups

### ISSUE THREAD

**Issue by ryvnf** [minetest/minetest#11120]
## Problem

Currently you can use `tool_capabilities` to change the digging speed of various tools. But there is no way to change the digging speed per player.

Setting the digging speed per player can have many uses. It can be used to apply penalties for digging times when players does things like swimming. It can also be used to create potions and special food which change the players digging speed. And be used to implement faster digging times for levels when a player levels up various skills like Mining in an RPG type game.

## Solutions

I propose to create two methods in player objects called `set_digtime_multiplier` and `get_digtime_multiplier`. The `set_digtime_multiplier` function takes in a table which contains a digging speed multiplier for different groups. So the following for example:

```
player:set_digtime_multiplier({
    "cracky": 0.8
    "crumbly": 1.1
})
```

Will make the player dig `cracky` nodes 20 % faster than normally and `crumbly` nodes 10 % slower than normally.

It should also be possible to specify the default dig speed multiplier using the empty string `""` as a key like the following.

```
player:set_digtime_multiplier({
    "": 1.1
    "cracky": 1.2
    "crumbly": 0.9
})
```

This gives the same result as the other example but also gives the player a 10 % digging speed bonus for all groups not included in the table.

The digging speed multiplier will only be in effect if the tool the player uses to dig has that group and the node the player digs has that group. Having a faster digging time with `cracky` will for example not help when digging stone with a shovel or when digging dirt with a pickaxe.

The `get_digtime_multiplier` does what you expect and returns the table set by `set_digtime_multiplier`.

**Comment by LoneWolfHT** [comment]
```
player:set_digtime_multiplier({
    "": 1.1
    "cracky": 1.2
    "crumbly": 0.9
})
```
Is this valid Lua?

**Comment by ryvnf** [comment]
Sorry. I was writing the wrong syntax for some reason. Updated now!

**Comment by AFCMS** [comment]
> ```
> player:set_digtime_multiplier({
>     ["cracky"] = 0.8,
>     ["crumbly"] = 1.1
> })
> ```

Useful feature! I really hope this will get implemented.

**Comment by fluxionary** [comment]
Doesn't the itemstack metadata tool_capabilities thing already support such behavior?

**Comment by SwissalpS** [comment]
doesn't that override the tools of all players though? OP is asking for per player overrides.

**Comment by AFCMS** [comment]
Overriding individual tools capabilities is already possible, but this doesn't allow for globally improving digging speed for a specific player (like the minecraft haste effect).

**Comment by fluxionary** [comment]
> doesn't that override the tools of all players though? 

no, just the one stack. 

> Overriding individual tools capabilities is already possible, but this doesn't allow for globally improving digging speed for a specific player (like the minecraft haste effect).

that can be done (to an extent) by overriding the toolcaps of the player's hand. (though you can't override the toolcaps of an empty stack effectively). i'm working on some things to manage this, though they're not done yet:
* https://github.com/fluxionary/minetest-hand_monoid
* https://github.com/fluxionary/minetest-hand_control

**Comment by AFCMS** [comment]
> that can be done (to an extent) by overriding the toolcaps of the player's hand. (though you can't override the toolcaps of an empty stack effectively).

But, this will only affect the empty hand, not any tools a player is holding, right?

**Comment by pandaro** [comment]
can be obtained via set_tool_capabilities()

**Comment by pandaro** [comment]
> 

This Is not true
For example, pmeta=player:get_meta()
Pmeta:set_string(digmulti,1.1)
Tool:setcapability(player:get_meta(digmulti)

Is Easy doable, then this issues can be closed

**Comment by fluxionary** [comment]
@pandaro yeah, but it's incredibly hacky to do it that way.  

you'd have to run that for every player on every server step, and you'd end up w/ lots of tools w/ random metadata, which will cause problems when using them w/ any mechanic which requires identical items (e.g. smartshops).

**Comment by pandaro** [comment]
> @pandaro yeah, but it's incredibly hacky to do it that way.
> 
> you'd have to run that for every player on every server step, and you'd end up w/ lots of tools w/ random metadata, which will cause problems when using them w/ any mechanic which requires identical items (e.g. smartshops).

It's true that it's a bit hacky, but it doesn't require checking every server step.
I'm probing for this type of usage through after_use(), so I can verify who's wielding the tool and apply the necessary enhancements to the tool capabilities.
I believe that a function capable of directly altering the excavation capabilities per player is partial (one would also be needed for the damage and for all the tool capabilities).
From my experiments, however, I believe that it would be more appropriate to introduce a function that is called before_use() without hindering the basic mechanics of a tool, which on_use() instead does.

**Comment by fluxionary** [comment]
> it doesn't require checking every server step.

true, but it does if you want any sort of responsiveness

> before_use()

hm personally, i feel like it'd be more useful to have callbacks like `on_wield_item` and `on_unwield_item`. but at that point, we're already adding new API, and might as well implement `set_digtime_multiplier`, as that more directly allows implementing the desired functionality. 

but this is all moot if someone doesn't implement it.

**Comment by pandaro** [comment]
> hm personally, i feel like it'd be more useful to have callbacks like `on_wield_item` and `on_unwield_item`. but at that point, we're already adding new API, and might as well implement `set_digtime_multiplier`, as that more directly allows implementing the desired functionality.

I fully agree, on wield and on_unwield are two excellent functions for loading and unloading tool_capabilities.

Instead I think that a function like set_digtime_multiplier() is usually incomplete, if you were to implement something like this then it would be much more convenient to aim for:
player:set_tool_capabilities_multiplier() able to directly create a multiplier for each parameter of a tool, and make the damage also have its multiplication parameter, as well as the tool_usage.

> but this is all moot if someone doesn't implement it.


In my eyes it is obvious that this type of call has an RPG like target.
Implementing an RPG requires a considerable collective effort, however, if the tools are not there, it is not feasible at all.
